geoips/interfaces/base.py

```diff
@@ -0,0 +1,360 @@
+    ``__call__`` from the plugin modules and using them in the objects.
+
+    For a module to be converted into an object it must meet the following requirements:
+    - The module must define a docstring. This will be used as the docstring for the
+      plugin class as well as the description for the plugin when requested on the
+      command line.
+    - The following global attributes must be defined in the module:
+      - name: The name of the plugin (must be unique for the interface)
+      - family: The family of plugins within the interface that the plugin belongs to
+      - description: A short description of the plugin. This will be used frequently on
+        the command line and should be limited to something small.
+    - A callable named `call` that will be called when the plugin is used.
+
+    Parameters
+    ----------
+    interface : str
+      Name of the interface that the desired plugin belongs to.
+    plugin : str
+      Name of the desired plugin.
+    module_call_func : str, optional
+      Use of this parameter is deprecated. Name of the callable within the plugin
+      (defaults to "call").
+    obj_attrs : dict, optional
+      Additional attributes to be assigned to the plugin object.
+
+    Returns
+    -------
+    An object of type ``<interface>InterfacePlugin`` where ``<interface>`` is the name
+    of the interface that the desired plugin belongs to.
+    """
+    obj_attrs["interface"] = interface
+    obj_attrs["module"] = module
+
+    try:
+        obj_attrs["__doc__"] = module.__doc__
+    except AttributeError:
+        if plugin_deprecations:
+            warn(
+                f"Plugin module '{module.__name__}' does not have a docstring. "
+                "Please add a module-level docstring",
+                DeprecationWarning,
+                stacklevel=1,
+            )
+        else:
+            raise PluginError(
+                f"Required '__doc__' attribute not found in '{module.__name__}'"
+            )
+
+    try:
+        obj_attrs["name"] = module.name
+    except AttributeError:
+        if plugin_deprecations:
+            warn(
+                f"Variable 'name' not found in plugin module '{module.__name__}'."
+                f"Assuming module name is "
+                f"`{'.'.join(module.__name__.split('.')[-2:])}`. "
+                "Please update to add 'name' variable."
+            )
+            obj_attrs["name"] = module.__name__.split(".")[-1]
+        else:
+            raise PluginError(
+                f"Required 'name' attribute not found in '{module.__name__}'"
+            )
+
+    try:
+        obj_attrs["family"] = module.family
+    except AttributeError:
+        if plugin_deprecations:
+            warn(
+                f"Use of the '{interface.deprecated_family_attr}' attribute is "
+                "deprecated. All uses should be replaced with an attribute named "
+                "'family'. In the future this will result in a TypeError.",
```

Comment:
@jsolbrig Should these all be "PluginError" vs "TypeError" in the warnings ?  And, do we care?

Reply:
It would probably be good to standardize here. Probably `PluginError` would be better. I probably added that after writing the deprecation method.

Maybe this can be done in an issue as a bug fix since it doesn't actually impact how the code runs?

Reply:
I just went ahead and changed it.